File: src/services/events.ts
import { listen } from '@tauri-apps/api/event';
import type { DownloadProgress, LiveTelemetryEvent, LiveTelemetryStatus, RuntimeSwitchResult, TrackedDownload, UpdateCheckResult } from '../types';

export type EventUnlisten = () => void;

/**
 * Own asynchronous Tauri listener registration for a React effect.
 *
 * `listen()` resolves asynchronously.  A normal effect cleanup can therefore
 * run before it yields its unlisten callback, leaking a stale listener.  Keep
 * the listener and the callback's liveness in one small scope so callers can
 * safely register several listeners without hand-rolled mutable variables.
 */
export interface AsyncListenerScope {
  register: (subscribe: () => Promise<EventUnlisten>) => void;
  dispose: () => void;
  isActive: () => boolean;
}

export function createAsyncListenerScope(onError?: (error: unknown) => void): AsyncListenerScope {
  let disposed = false;
  const unlisteners = new Set<EventUnlisten>();

  const dispose = () => {
    if (disposed) return;
    disposed = true;
    for (const unlisten of unlisteners) {
      unlisten();
    }
    unlisteners.clear();
  };

  return {
    register: (subscribe) => {
      void subscribe()
        .then((unlisten) => {
          if (disposed) {
            unlisten();
            return;
          }
          unlisteners.add(unlisten);
        })
        .catch((error: unknown) => onError?.(error));
    },
    dispose,
    isActive: () => !disposed,
  };
}

export interface ProgressEvent {
  downloadId: string;
  progress: DownloadProgress;
}

export interface CompleteEvent {
  downloadId: string;
  manifestId?: string;
}

export interface ErrorEvent {
  downloadId: string;
  error: string;
}

export interface AuthWaitingEvent {
  downloadId: string;
  message: string;
}

export interface AuthSuccessEvent {
  downloadId: string;
}

export interface AuthErrorEvent {
  downloadId: string;
  error: string;
}

export interface SteamAuthQrLineEvent {
  line: string;
}

export interface MelonLoaderInstallingEvent {
  downloadId: string;
  environmentId: string;
  message: string;
}

export interface MelonLoaderInstalledEvent {
  downloadId: string;
  environmentId: string;
  message: string;
  version?: string;
}

export interface MelonLoaderErrorEvent {
  downloadId: string;
  environmentId: string;
  message: string;
}

export interface UpdateAvailableEvent {
  environmentId: string;
  updateResult: UpdateCheckResult;
}

export interface UpdateCheckCompleteEvent {
  environmentId: string;
  updateResult: UpdateCheckResult;
}

export type RuntimeSwitchEvent = RuntimeSwitchResult;

export interface ModsChangedEvent {
  environmentId: string;
}

export interface ModsSnapshotUpdatedEvent {
  environmentId: string;
  snapshot: {
    mods: Array<{
      name: string;
      fileName: string;
      path: string;
      version?: string;
      source?: string;
      sourceUrl?: string;
      disabled?: boolean;
      modStorageId?: string;
      managed?: boolean;
    }>;
    modsDirectory: string;
    count: number;
  };
}

export interface PluginsChangedEvent {
  environmentId: string;
}

export interface UserLibsChangedEvent {
  environmentId: string;
}

export interface ModUpdatesCheckedEvent {
  environmentId: string;
  count: number;
  updates: Array<{
    modFileName: string;
    modName: string;
    currentVersion: string;
    latestVersion: string;
    source: string;
  }>;
}

export interface ModMetadataRefreshStatusEvent {
  activeCount: number;
  running: boolean;
}

export type TrackedDownloadUpdatedEvent = TrackedDownload;

export async function onLiveTelemetryEvent(handler: (data: LiveTelemetryEvent) => void): Promise<() => void> {
  return await listen<LiveTelemetryEvent>('live_telemetry_event', (event) => handler(event.payload));
}

export async function onLiveTelemetryStatus(handler: (data: LiveTelemetryStatus) => void): Promise<() => void> {
  return await listen<LiveTelemetryStatus>('live_telemetry_status', (event) => handler(event.payload));
}

export async function onProgress(handler: (data: DownloadProgress) => void): Promise<() => void> {
  return await listen<ProgressEvent>('download_progress', (event) => {
    handler(event.payload.progress);
  });
}

export async function onComplete(handler: (data: CompleteEvent) => void): Promise<() => void> {
  return await listen<CompleteEvent>('download_complete', (event) => {
    handler(event.payload);
  });
}

export async function onError(handler: (data: ErrorEvent) => void): Promise<() => void> {
  return await listen<ErrorEvent>('download_error', (event) => {
    handler(event.payload);
  });
}

export async function onAuthWaiting(handler: (data: AuthWaitingEvent) => void): Promise<() => void> {
  return await listen<AuthWaitingEvent>('auth_waiting', (event) => {
    handler(event.payload);
  });
}

export async function onAuthSuccess(handler: (data: AuthSuccessEvent) => void): Promise<() => void> {
  return await listen<AuthSuccessEvent>('auth_success', (event) => {
    handler(event.payload);
  });
}

export async function onAuthError(handler: (data: AuthErrorEvent) => void): Promise<() => void> {
  return await listen<AuthErrorEvent>('auth_error', (event) => {
    handler(event.payload);
  });
}

export async function onSteamAuthQrLine(handler: (data: SteamAuthQrLineEvent) => void): Promise<() => void> {
  return await listen<SteamAuthQrLineEvent>('steam_auth_qr_line', (event) => {
    handler(event.payload);
  });
}

export async function onMelonLoaderInstalling(handler: (data: MelonLoaderInstallingEvent) => void): Promise<() => void> {
  return await listen<MelonLoaderInstallingEvent>('melonloader_installing', (event) => {
    handler(event.payload);
  });
}

export async function onMelonLoaderInstalled(handler: (data: MelonLoaderInstalledEvent) => void): Promise<() => void> {
  return await listen<MelonLoaderInstalledEvent>('melonloader_installed', (event) => {
    handler(event.payload);
  });
}

export async function onMelonLoaderError(handler: (data: MelonLoaderErrorEvent) => void): Promise<() => void> {
  return await listen<MelonLoaderErrorEvent>('melonloader_error', (event) => {
    handler(event.payload);
  });
}

export async function onUpdateAvailable(handler: (data: UpdateAvailableEvent) => void): Promise<() => void> {
  return await listen<UpdateAvailableEvent>('update_available', (event) => {
    handler(event.payload);
  });
}

export async function onUpdateCheckComplete(handler: (data: UpdateCheckCompleteEvent) => void): Promise<() => void> {
  return await listen<UpdateCheckCompleteEvent>('update_check_complete', (event) => {
    handler(event.payload);
  });
}

export async function onRuntimeSwitch(handler: (data: RuntimeSwitchEvent) => void): Promise<() => void> {
  return await listen<RuntimeSwitchEvent>('steam_runtime_switched', (event) => {
    handler(event.payload);
  });
}

export async function onModsChanged(handler: (data: ModsChangedEvent) => void): Promise<() => void> {
  return await listen<ModsChangedEvent>('mods_changed', (event) => {
    handler(event.payload);
  });
}

export async function onModsSnapshotUpdated(handler: (data: ModsSnapshotUpdatedEvent) => void): Promise<() => void> {
  return await listen<ModsSnapshotUpdatedEvent>('mods_snapshot_updated', (event) => {
    handler(event.payload);
  });
}

export async function onPluginsChanged(handler: (data: PluginsChangedEvent) => void): Promise<() => void> {
  return await listen<PluginsChangedEvent>('plugins_changed', (event) => {
    handler(event.payload);
  });
}

export async function onUserLibsChanged(handler: (data: UserLibsChangedEvent) => void): Promise<() => void> {
  return await listen<UserLibsChangedEvent>('userlibs_changed', (event) => {
    handler(event.payload);
  });
}

export async function onModUpdatesChecked(handler: (data: ModUpdatesCheckedEvent) => void): Promise<() => void> {
  return await listen<ModUpdatesCheckedEvent>('mod_updates_checked', (event) => {
    handler(event.payload);
  });
}

export async function onModMetadataRefreshStatus(
  handler: (data: ModMetadataRefreshStatusEvent) => void
): Promise<() => void> {
  return await listen<ModMetadataRefreshStatusEvent>('mod_metadata_refresh_status', (event) => {
    handler(event.payload);
  });
}

export async function onTrackedDownloadUpdated(
  handler: (data: TrackedDownloadUpdatedEvent) => void
): Promise<() => void> {
  return await listen<TrackedDownloadUpdatedEvent>('tracked_download_updated', (event) => {
    handler(event.payload);
  });
}

File: src-tauri/src/events.rs
use crate::types::{DownloadProgress, TrackedDownload, UpdateCheckResult};
use tauri::{AppHandle, Emitter, Runtime};

pub fn emit_progress<R: Runtime>(
    app: &AppHandle<R>,
    progress: DownloadProgress,
) -> Result<(), tauri::Error> {
    app.emit(
        "download_progress",
        serde_json::json!({
            "downloadId": progress.download_id.clone(),
            "progress": progress
        }),
    )
}

pub fn emit_complete<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
    manifest_id: Option<String>,
) -> Result<(), tauri::Error> {
    app.emit(
        "download_complete",
        serde_json::json!({
            "downloadId": download_id,
            "manifestId": manifest_id
        }),
    )
}

pub fn emit_error<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
    error: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "download_error",
        serde_json::json!({ "downloadId": download_id, "error": error }),
    )
}

pub fn emit_auth_waiting<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
    message: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "auth_waiting",
        serde_json::json!({ "downloadId": download_id, "message": message }),
    )
}

pub fn emit_auth_success<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "auth_success",
        serde_json::json!({ "downloadId": download_id }),
    )
}

pub fn emit_auth_error<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
    error: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "auth_error",
        serde_json::json!({ "downloadId": download_id, "error": error }),
    )
}

pub fn emit_melonloader_installing<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
    message: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "melonloader_installing",
        serde_json::json!({ "downloadId": download_id, "message": message }),
    )
}

pub fn emit_melonloader_installed<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
    message: String,
    version: Option<String>,
) -> Result<(), tauri::Error> {
    app.emit(
        "melonloader_installed",
        serde_json::json!({
            "downloadId": download_id,
            "message": message,
            "version": version
        }),
    )
}

pub fn emit_melonloader_error<R: Runtime>(
    app: &AppHandle<R>,
    download_id: String,
    message: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "melonloader_error",
        serde_json::json!({ "downloadId": download_id, "message": message }),
    )
}

pub fn emit_update_available<R: Runtime>(
    app: &AppHandle<R>,
    environment_id: String,
    update_result: UpdateCheckResult,
) -> Result<(), tauri::Error> {
    app.emit(
        "update_available",
        serde_json::json!({
            "environmentId": environment_id,
            "updateResult": update_result
        }),
    )
}

pub fn emit_update_check_complete<R: Runtime>(
    app: &AppHandle<R>,
    environment_id: String,
    update_result: UpdateCheckResult,
) -> Result<(), tauri::Error> {
    app.emit(
        "update_check_complete",
        serde_json::json!({
            "environmentId": environment_id,
            "updateResult": update_result
        }),
    )
}

pub fn emit_mods_changed<R: Runtime>(
    app: &AppHandle<R>,
    environment_id: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "mods_changed",
        serde_json::json!({ "environmentId": environment_id }),
    )
}

pub fn emit_mods_snapshot_updated<R: Runtime>(
    app: &AppHandle<R>,
    environment_id: String,
    snapshot: serde_json::Value,
) -> Result<(), tauri::Error> {
    app.emit(
        "mods_snapshot_updated",
        serde_json::json!({
            "environmentId": environment_id,
            "snapshot": snapshot
        }),
    )
}

pub fn emit_plugins_changed<R: Runtime>(
    app: &AppHandle<R>,
    environment_id: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "plugins_changed",
        serde_json::json!({ "environmentId": environment_id }),
    )
}

pub fn emit_userlibs_changed<R: Runtime>(
    app: &AppHandle<R>,
    environment_id: String,
) -> Result<(), tauri::Error> {
    app.emit(
        "userlibs_changed",
        serde_json::json!({ "environmentId": environment_id }),
    )
}

pub fn emit_mod_updates_checked<R: Runtime>(
    app: &AppHandle<R>,
    environment_id: String,
    count: usize,
    updates: Vec<serde_json::Value>,
) -> Result<(), tauri::Error> {
    app.emit(
        "mod_updates_checked",
        serde_json::json!({
            "environmentId": environment_id,
            "count": count,
            "updates": updates
        }),
    )
}

pub fn emit_mod_metadata_refresh_status<R: Runtime>(
    app: &AppHandle<R>,
    active_count: usize,
) -> Result<(), tauri::Error> {
    app.emit(
        "mod_metadata_refresh_status",
        serde_json::json!({
            "activeCount": active_count,
            "running": active_count > 0
        }),
    )
}

pub fn emit_tracked_download_updated<R: Runtime>(
    app: &AppHandle<R>,
    download: TrackedDownload,
) -> Result<(), tauri::Error> {
    app.emit("tracked_download_updated", download)
}
